File: interleaved/interleaf.cpp
#include <iostream>
#include <string>
using namespace std;

bool is_interleaf(string s1, string s2, string s3);
int main(){
	string s1 = "abb";
	string s2 = "baa";
	string s3 = "abbbaa";
	string s4 = "ababba";
	bool res1 = is_interleaf(s1,s2,s3);
	bool res2 = is_interleaf(s1,s2,s4);
	cout << "res1: " << res1 << endl;
	cout << "res2: " << res2 << endl;
	return 0;
}
bool is_interleaf(string s1, string s2, string s3){
	if (s1.size() == 0 and s2.size() == 0)
		return true;
	if (s1.size() > 0 && s1[0] == s3[0])
		return is_interleaf(s1.substr(1, s1.size()-1),s2,s3.substr(1, s3.size()-1));
	if (s2.size() > 0 && s2[0] == s3[0])
		return is_interleaf(s1,s2.substr(1, s2.size()-1),s3.substr(1, s3.size()-1));
	return false;
}

File: interleaved/interleaf.py
def interleave(s1,s2,s3):
    if len(s1) == 0 and len(s2) == 0:
        return True
    if len(s1) > 0 and s1[0] == s3[0]:
        return interleave(s1[1:],s2,s3[1:])
    if len(s2) > 0 and s2[0] == s3[0]:
        return interleave(s1,s2[1:],s3[1:])
    return False

s1 = "aab"
s2 = "cd"
s3 = "aacdb"
print(interleave(s1,s2,s3))

s1 = "abb"
s2 = "baa"
s3 = "abbbaa"
s4 = "ababba"
print(interleave(s1,s2,s3))
print(interleave(s1,s2,s4))
